File: src/employee-policy-assistant/Policy.Harness/Bootstrap.cs
// SPDX-License-Identifier: Apache-2.0
using Ioka.Munarium.Client;
using Policy.Core;

namespace Policy.Harness;

public static class Bootstrap
{
    public const string ClientRevision = "bb6e92a72a3944cff4d4bf0c1b470afcf3f4dfb3";
    public static string Endpoint => Environment.GetEnvironmentVariable("MUNARIUM_REST_URL") ?? "http://server:8080";
    public static MunariumClient Ops(bool management = false) => MunariumClient.Rest(new() { Endpoint = Endpoint, Token = Environment.GetEnvironmentVariable(management ? "MUNARIUM_MGMT_TOKEN" : "MUNARIUM_TOKEN") ?? throw new InvalidOperationException("Bootstrap credential absent."), Uid = "policy-bootstrap" });
    public static string Model(string provider) => provider switch { "fixture" => "policy-selected", _ => Environment.GetEnvironmentVariable(provider.ToUpperInvariant() + "_MODEL") ?? throw new InvalidOperationException("Preferred model missing.") };
    public static async Task Run(string provider, string inputs, string credentials, string work, bool approve)
    {
        if (provider is not ("fixture" or "openai" or "anthropic" or "openrouter")) throw new ArgumentException("Unsupported demo provider.");
        Fixtures.Verify(inputs);
        await using var ops = Ops();
        for (var attempt = 0; ; attempt++)
        {
            try { var version = await ops.ServerVersionAsync(); if (version.Version != "1.1.1") throw new InvalidDataException("Server 1.1.1 required."); break; }
            catch (MunariumException) when (attempt < 60) { await Task.Delay(2000); }
        }
        var template = File.ReadAllText("runbooks/policy.yaml");
        var shape = File.ReadAllText("shapes/documents.yaml");
        var model = Model(provider);
        var revision = Storage.Hash(File.ReadAllText(Path.Combine(inputs, "manifest.json")) + template + shape + provider + model)[..12];
        var name = "policy-" + revision;
        var config = name + "-model";
        var connection = provider == "fixture" ? "endpoint: http://provider-fixture:11434" : $"credentialRef: {{ env: {provider.ToUpperInvariant()}_API_KEY }}";
        await ops.Providers.ApplyConfigAsync($$"""
            apiVersion: munarium.ioka.io/v1
            kind: ProviderConfig
            metadata: { name: {{config}} }
            spec:
              provider: {{(provider == "fixture" ? "ollama" : provider)}}
              {{connection}}
              models:
                complete: [{{model}}]
                fast: {{model}}
                capable: {{model}}
              budgets: { rpm: 60, dailyTokens: { fast: 200000, capable: 200000 } }
            """);
        if (!(await ops.Providers.HealthAsync(config)).Healthy) throw new InvalidOperationException("Provider health failed.");
        await ops.Runbooks.ApplyShapeAsync(shape);
        await ops.Runbooks.ApplyRunbookAsync(template.Replace("__NAME__", name).Replace("__PROVIDER__", config));
        foreach (var path in Directory.GetFiles(inputs, "*.txt", SearchOption.AllDirectories).Order(StringComparer.Ordinal))
        {
            var relative = Path.GetRelativePath(inputs, path).Replace('\\', '/');
            var result = await ops.Ingest.IngestAsync(new() { Filename = name + "/" + relative, MediaType = "text/plain", ContentBase64 = Convert.ToBase64String(File.ReadAllBytes(path)) });
            if (result.Error is not null || !result.BoundTo.Contains(name + "-" + relative.Split('/')[0])) throw new InvalidDataException("Ingest binding failed.");
        }
        var run = await ops.Runbooks.RunRunbookAsync(name);
        Storage.Save(Path.Combine(work, "bootstrap-" + provider + "-" + run.RunId + ".json"), new { run.RunId, name, revision, provider, model, clientRevision = ClientRevision, manifest = Storage.Read<object>(Path.Combine(inputs, "manifest.json")) });
        if (!approve) throw new InvalidOperationException("Inspect the saved run; rerun bootstrap with --approve for this isolated test corpus.");
        for (var step = 0; step < 8; step++)
        {
            var status = await ops.Runbooks.GetRunAsync(run.RunId);
            if (status.State == "done") break;
            var pending = status.Steps.Where(s => s.State == "awaiting_approval").ToArray();
            if (pending.Length != 1 || !new[] { "public", "west", "east", "hr" }.Any(c => pending[0].Name == "cutover:" + name + "-" + c)) throw new InvalidOperationException("Unexpected cutover; inspect saved run.");
            await ops.Runbooks.ApproveStepAsync(run.RunId, pending[0].Ordinal);
        }
        if ((await ops.Runbooks.GetRunAsync(run.RunId)).State != "done") throw new InvalidOperationException("Index run did not complete.");
        await using var issuer = Ops(true);
        foreach (var identity in new[] { "employee", "west", "east", "hr" })
        {
            var level = identity == "hr" ? 2 : identity == "employee" ? 0 : 1;
            var issued = await issuer.Tokens.MintAsync("policy-" + identity, level, ["query"], identity == "employee" ? [] : [identity], [name], 3600);
            Storage.Save(Path.Combine(credentials, identity + ".json"), new IdentityGrant(identity, "policy-" + identity, issued.Token, issued.ExpiresAt, name + "@1", revision, [new(config, provider == "fixture" ? "ollama" : provider, model)]));
        }
        Console.WriteLine($"Bootstrapped {provider}/{model}; four scoped identities; approved run {run.RunId}.");
    }
}

File: src/engineering-change-review/src/bootstrap.rs
// SPDX-License-Identifier: Apache-2.0
use crate::data::*;
use anyhow::{ensure, Result};
use base64::{engine::general_purpose::STANDARD, Engine};
use munarium_client::{dto, MunariumClient, MunariumClientOptions};
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::{collections::BTreeMap, env, fs, time::Duration};
#[derive(Clone, Serialize, Deserialize)]
pub struct Grant {
    pub token: String,
    pub uid: String,
    pub provider: String,
    pub model: String,
    pub config: String,
    pub scopes: BTreeMap<String, String>,
}
pub fn endpoint() -> String {
    env::var("MUNARIUM_REST_URL").unwrap_or("http://server:8080".into())
}
pub fn client_at(url: &str, token: &str, uid: &str) -> Result<MunariumClient> {
    Ok(MunariumClient::rest(
        MunariumClientOptions::new(url)
            .token(token)
            .uid(uid)
            .read_retries(0),
    )?)
}
pub fn ops(management: bool) -> Result<MunariumClient> {
    client_at(
        &endpoint(),
        &env::var(if management {
            "MUNARIUM_MGMT_TOKEN"
        } else {
            "MUNARIUM_TOKEN"
        })?,
        "engineering-bootstrap",
    )
}
pub fn grant() -> Result<Grant> {
    Ok(serde_json::from_value(read("/credentials/query.json")?)?)
}
pub async fn ready(c: &MunariumClient) -> Result<()> {
    for _ in 0..60 {
        if let Ok(v) = c.server_version().await {
            ensure!(v.version == "1.1.1", "Server 1.1.1 required");
            return Ok(());
        }
        tokio::time::sleep(Duration::from_secs(1)).await;
    }
    anyhow::bail!("Server not ready")
}
pub async fn mint(uid: &str, refs: Vec<String>, ttl: u64) -> Result<String> {
    Ok(ops(true)?
        .tokens
        .mint(dto::IssueTokenRequest {
            uid: uid.into(),
            access_level: 0,
            compartments: vec![],
            scopes: vec!["query".into()],
            runbook_refs: Some(refs),
            ttl_secs: Some(ttl),
        })
        .await?
        .token)
}
pub async fn run(provider: &str, approve: bool) -> Result<()> {
    ensure!(
        approve,
        "Explicit --approve required for isolated verified cutovers"
    );
    assignments(provider)?;
    verify("/inputs")?;
    let model = if provider == "fixture" {
        "engineering-fixture".into()
    } else {
        env::var(format!("{}_MODEL", provider.to_uppercase()))?
    };
    let template = fs::read_to_string("/app/runbooks/review.yaml")?;
    let shape = fs::read_to_string("/app/shapes/documents.yaml")?;
    let namespace = format!(
        "engineering-{}",
        &hash(format!(
            "{}{}{}{}{}",
            fs::read_to_string("/inputs/manifest.json")?,
            template,
            shape,
            provider,
            model
        ))[..12]
    );
    let config = format!("{namespace}-model");
    let api = ops(false)?;
    ready(&api).await?;
    let connection = if provider == "fixture" {
        "endpoint: http://provider-fixture:11434".into()
    } else {
        format!(
            "credentialRef: {{env: {}_API_KEY}}",
            provider.to_uppercase()
        )
    };
    let family = if provider == "fixture" {
        "ollama"
    } else {
        provider
    };
    let encoded = serde_json::to_string(&model)?;
    api.providers.apply_config(&format!("apiVersion: munarium.ioka.io/v1\nkind: ProviderConfig\nmetadata: {{name: {config}}}\nspec:\n  provider: {family}\n  {connection}\n  models: {{complete: [{encoded}], fast: {encoded}}}\n  budgets: {{rpm: 60, dailyTokens: {{fast: 200000}}}}\n")).await?;
    ensure!(
        api.providers.health(&config).await?.healthy,
        "Provider health failed"
    );
    api.runbooks.apply_shape(&shape, None).await?;
    let mut scopes = BTreeMap::new();
    for component in ["billing", "catalog", "security-admin"] {
        let name = format!("{namespace}-{component}");
        let book=template.replace("__NAME__",&name).replace("__PROVIDER__",&config).replace("__LEVEL__",if component=="security-admin"{"2"}else{"0"})+"\n  steps:\n    - resolveSources: {}\n    - buildIndex: {}\n    - verify: {}\n    - cutover: {approval: required}\n";
        api.runbooks.apply_runbook(&book).await?;
        for kind in ["architecture", "release"] {
            let content = if component == "security-admin" {
                b"RESTRICTED_ENGINEERING_SENTINEL. Fictional privileged procedure.".to_vec()
            } else {
                fs::read(format!("/inputs/documents/{component}-{kind}.txt"))?
            };
            let result = api
                .ingest
                .ingest(dto::IngestFileRequest {
                    filename: format!("{name}/{kind}.txt"),
                    media_type: "text/plain".into(),
                    sha256: Some(hash(&content)),
                    content_base64: STANDARD.encode(content),
                    collections: None,
                })
                .await?;
            ensure!(
                result.error.is_none() && result.bound_to == vec![name.clone()],
                "Unexpected ingest binding"
            );
        }
        let run = api.runbooks.run_runbook(&name, None).await?;
        save(
            format!("/work/bootstrap/{}.json", run.run_id),
            &json!({"run":run,"provider":provider,"model":model,"client_revision":REVISION,"component":component}),
        )?;
        let status = api.runbooks.get_run(&run.run_id).await?;
        let pending: Vec<_> = status
            .steps
            .iter()
            .filter(|s| s.state == "awaiting_approval")
            .collect();
        ensure!(
            pending.len() == 1 && pending[0].name == format!("cutover:{name}"),
            "Unexpected cutover"
        );
        api.runbooks
            .approve_step(&run.run_id, pending[0].ordinal)
            .await?;
        ensure!(
            api.runbooks.get_run(&run.run_id).await?.state == "done",
            "Incomplete index run"
        );
        if component != "security-admin" {
            scopes.insert(component.into(), format!("{name}@1"));
        }
    }
    let token = mint(
        "engineering-ci",
        scopes
            .values()
            .map(|r| r.split('@').next().unwrap().into())
            .collect(),
        3600,
    )
    .await?;
    save(
        "/credentials/query.json",
        &Grant {
            token,
            uid: "engineering-ci".into(),
            provider: provider.into(),
            model,
            config,
            scopes,
        },
    )?;
    Ok(())
}
